Allow easier debugging of Action Dispatch requests

dorianmariecom opened the pull request:
Kernel#method was redefined so one couldn't do for instance.

method(:POST).source_location

Now when called without arguments it returns the method of the
request and when called with arguments it uses Kernel#method

Which makes debugging easier

### Other Information

I was trying to debug ActionDispatch::Request's methods and first thought I was doing something wrong when calling `method(:POST)`. I first did an `alias_method :_method, :method` but thought this fix could be more intuitive and helpful for others too.

dorianmariefr commented:
failure seems unrelated:

```
Failure:
ActiveRecord::ConnectionAdapters::ConnectionPoolThreadTest#test_disconnect_and_clear_reloadable_connections_are_able_to_preempt_other_waiting_threads [/rails/activerecord/test/cases/connection_pool_test.rb:665]:
disconnect is not able to preempt other waiting threads
```

casperisfine commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -195,8 +195,15 @@ def request_method_symbol
+    # For debugging purposes, when called with arguments this method will
+    # fallback to Kernel#method
+    def method(*args)
+      if args.any?
```
`any?` isn't a good substitute for `!empty?`:

```ruby
>> [false].any?
=> false
```

dorianmariefr replied:
didn't know that, thanks

casperisfine commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -195,8 +195,15 @@ def request_method_symbol
+    #
+    # For debugging purposes, when called with arguments this method will
+    # fallback to Kernel#method
+    def method(*args)
```
`Object#method` current signature is just `method(sym)`, but when delegating to a method we have no control over, it's best to not assume the signature. So we'd need `ruby2_keywords` here.

casperisfine commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -195,8 +195,15 @@ def request_method_symbol
+    # fallback to Kernel#method
+    def method(*args)
+      if args.any?
+        super(*args)
```
```suggestion
        super
```

casperisfine commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -195,8 +195,15 @@ def request_method_symbol
-      @method ||= check_method(get_header("rack.methodoverride.original_method") || get_header("REQUEST_METHOD"))
+    #
+    # For debugging purposes, when called with arguments this method will
+    # fallback to Kernel#method
```
```suggestion
    # fallback to Object#method
```

Commit:
Allow easier debugging of Action Dispatch requests

Kernel#method was redefined so one couldn't do for instance.

method(:POST).source_location

Now when called without arguments it returns the method of the
request and when called with arguments it uses Kernel#method

Which makes debugging easier

Co-authored-by: Joé Dupuis <joe@dupuis.io>

dorianmariefr commented:
Thanks @casperisfine for the merge, too bad I didn't meet you at railsconf I wanted to say thanks for looking at my pull requests

byroot commented:
Maybe next time!
